pyrage-stubs/pyrage-stubs/pyrage.pyi

```diff
@@ -1,9 +1,13 @@
+from io import BufferedIOBase
 from typing import Sequence, Union
 
 from . import ssh, x25519
 
 Identity = Union[ssh.Identity, x25519.Identity]
 Recipient = Union[ssh.Recipient, x25519.Recipient]
+FileLikeOrPath = Union[str, BufferedIOBase]
 
 def encrypt(plaintext: bytes, recipients: Sequence[Recipient]) -> bytes: ...
+def encrypt_file(infile: FileLikeOrPath, outfile: FileLikeOrPath, recipients: Sequence[Recipient]) -> None: ...
```

Comment:
As a matter of personal preference: I'm not a huge fan of APIs that can take either paths *or* pre-opened files, since they have confusing ownership behavior (who owns the underlying FD?) and don't tell the user which one they should prefer.

It looks like `pyo3-file` papers over that first concern, but I think I'd prefer that we drop that dependency entirely and settle on either a path-based API or a file-like API. My slight preference would be for the latter, but I'd also be okay with the former if you think it'll be too annoying (since pyo3 doesn't have "native" support for file-like objects).

After that, the API here (and below) should be:

```suggestion
def encrypt_file(infile: BinaryIO, outfile: BinaryIO, recipients: Sequence[Recipient]) -> None: ...
```

(or maybe a custom protocol type instead of `BinaryIO`, if needed.)

Reply:
First of all thank you for taking a look at this!

I did notice during testing that there seems to be quite a large performance difference between opening files in Python and opening them in Rust. For my intentionally large test file (25GB), the difference was almost a factor of 2.
Thus from a performance-perspective (and implementation complexity) I'd be slightly in favor of having opening the files in Rust and not Python.
The `pyo3-file` dependency is orthogonal to the issue of choosing either a path or a file: It simply implements `Read` and `Write` for the Python object.
However if you'd prefer to have fewer dependencies, these traits can also be implemented inside pyrage.

I don't have a strong preference for either a path based or a file based API and since this is your library, I'll go with whatever you think makes the most sense.
A third option would also be to supply both (e.g. `encrypt_file` and `encrypt_file_at`).
Let me know which option you prefer and I'll adapt the PR.

Reply:
No problem! Thank you for going through the review process with me 🙂 

> I did notice during testing that there seems to be quite a large performance difference between opening files in Python and opening them in Rust. For my intentionally large test file (25GB), the difference was almost a factor of 2.

Yeah, that's true -- interacting through a file-like object is going to be a lot slower on each `read` invocation, since we chase through the object itself and have to deal with the GIL and interpreter overhead.

Based on that, let's start with just a path-based API. That will be less generic than a file-like API, but IMO that's fine until/unless more users come forward and explicitly request file-like object support.

Reply:
(And yep, let's drop the `pyo3-file` dep for now -- if I'm understanding correctly, we won't need its trait impls now.)